### ISSUE TITLE
Accessing the currently inherited/included/extended type in macro hooks

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#10710]
Continued from https://github.com/crystal-lang/crystal/issues/9488#issuecomment-839950518.

The `@type` metavariable in an `inherited` / `included` / `extended` macro hook refers to the type that triggers the macro hook, not the macro's owner. That is accessible through `@type.ancestors.first`, but apart from not expressing the intent clearly, this is error-prone for two reasons:

* Since `@type` is never a metaclass, the `extended` hook should actually use `@type.class_type.ancestors.first` instead.
* Prepended modules (#10504) can change the semantics of `TypeNode#ancestors`, as these modules should always appear before included ones.

For a concrete use case, suppose that `String`'s [final hook](https://github.com/crystal-lang/crystal/blob/80c5b469a91a6481c2797df30fa422f2d6650e37/src/string.cr#L144) is to be extracted into a macro. This can be done by:

```crystal
macro final!
  macro inherited
    \{% raise "Cannot inherit from #{@type.ancestors.first}" %}
  end

  # ditto for `included` and `extended`
end

class String
  final!
end
```

I propose that one of the following ways be added to access such an owner type:

* A new metavariable called `@owner`:

  ```crystal
  macro inherited
    \{% raise "Cannot inherit from #{@owner}" %}
  end
  ```

* `self`: (I personally don't like this very much)

  ```crystal
  macro inherited
    \{% raise "Cannot inherit from #{self}" %}
  end
  ```

* If the macro hook may accept one positional argument, that argument is the owner type, otherwise parameter-less hooks continue to work as before:

  ```crystal
  macro inherited(owner)
    \{% raise "Cannot inherit from #{owner}" %}
  end

  macro included
    # okay, owner is inaccessible
  end
  ```

**Comment by straight-shoota** [comment]
I agree on not liking `self`. I would've initially preferred an argument, but considering it's a different semantic from Ruby's (basically self and other are switched), I'd probably prefer a macro variable `@owner`.
I can imagine this would even be useful for any macro method, not just hooks.
